File: day07/src/test/java/de/havox_design/aoc2015/day07/Day07Test.java
package de.havox_design.aoc2015.day07;

import java.util.stream.Stream;

import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.Arguments;
import org.junit.jupiter.params.provider.MethodSource;

class Day07Test {

    @SuppressWarnings("java:S2699")
    @Test
    void testMainClass() {
        MainClass.main(new String[0]);
    }

    @ParameterizedTest
    @MethodSource ("getDataForPart1")
    void testPart1(String fileName, int expectedNumberOfLights) {
        Assertions.assertEquals(expectedNumberOfLights, LogicGates.solvePart1(fileName));
    }

    private static Stream<Arguments> getDataForPart1() {
        return Stream.of(
            Arguments.of("part1sample1.txt", 72),
            Arguments.of("part1sample2.txt", 507),
            Arguments.of("part1sample3.txt", 492),
            Arguments.of("part1sample4.txt", 114),
            Arguments.of("part1sample5.txt", 65412),
            Arguments.of("part1sample6.txt", 65079),
            Arguments.of("part1sample7.txt", 123),
            Arguments.of("part1sample8.txt", 456)
        );
    }
}

File: day07/src/test/kotlin/de/havox_design/aoc2015/day07/Day07Test.kt
package de.havox_design.aoc2015.day07

import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.Arguments
import org.junit.jupiter.params.provider.MethodSource
import java.util.stream.Stream

class Day07Test {
    @Test
    fun testMainClass() {
        MainClass.main(arrayOf())
    }

    @ParameterizedTest
    @MethodSource("getDataForTestProcessPart1")
    fun testProcessPart1(filename: String, expectedResult: Int) =
        LogicGates(filename).processPart1().shouldBe(expectedResult)

    @ParameterizedTest
    @MethodSource("getDataForTestProcessPart2")
    fun testProcessPart2(filename: String, expectedResult: Int) =
        LogicGates(filename).processPart2().shouldBe(expectedResult)

    companion object {
        @JvmStatic
        private fun getDataForTestProcessPart1(): Stream<Arguments> =
            Stream.of(
                Arguments.of("part1sample1.txt", 72),
                Arguments.of("part1sample2.txt", 507),
                Arguments.of("part1sample3.txt", 492),
                Arguments.of("part1sample4.txt", 114),
                Arguments.of("part1sample5.txt", 65412),
                Arguments.of("part1sample6.txt", 65079),
                Arguments.of("part1sample7.txt", 123),
                Arguments.of("part1sample8.txt", 456)
            )

        @JvmStatic
        private fun getDataForTestProcessPart2(): Stream<Arguments> =
            Stream.of(
                Arguments.of("sample.txt", 42)
            )
    }
}

private fun Int.shouldBe(expectation: Int) = Assertions.assertEquals(expectation, this)
